File: Programmers/전력망나누기.py
from collections import deque

def solution(n, wires):
    answer = 101
    graph = [[] for _ in range(n+1)]
    #make graph
    for wire in wires:
        graph[wire[0]].append(wire[1])
        graph[wire[1]].append(wire[0])
    
    for wire in wires:
        checked = [False]*(n+1) 
        cnt = 1
        
        #BFS
        q = deque()
        q.append(wire[0])
        checked[wire[0]] = True
        checked[wire[1]] = True
        
        while q:
            cur_node = q.popleft()
            
            for next_node in graph[cur_node]:
                if not checked[next_node]:
                    checked[next_node] = True
                    q.append(next_node)
                    cnt += 1
                    
        dif = abs((n - cnt) - cnt)
        
        if dif < answer:
            answer = dif
    
    return answer

File: Programmers/전력망나누기.kt
import kotlin.math.*

class Solution {
    fun solution(n: Int, wires: Array<IntArray>): Int {
        var answer: Int = n
        var tree = Array(n+1,{mutableSetOf<Int>()}) 
        
        for (wire in wires){
            tree[wire[0]].add(wire[1])
            tree[wire[1]].add(wire[0])
        }
        
        for (wire in wires){
            tree[wire[0]].remove(wire[1])
            tree[wire[1]].remove(wire[0])
            
            var visited_cnt = DFS(wire[0],tree)
            var result = abs(n - visited_cnt - visited_cnt)
            
            if (result < answer){
                answer = result
            }
            
            tree[wire[0]].add(wire[1])
            tree[wire[1]].add(wire[0])
            
        }
        
        return answer
    }
    
    
    fun DFS(start:Int, tree:Array<MutableSet<Int>>):Int{
        var stack = mutableListOf<Int>(start)
        var checked = mutableSetOf<Int>(start)
        
        while(stack.isNotEmpty()){
            var cur_node = stack.removeAt(stack.size-1)
            
            for(next_node in tree[cur_node]){
                if (!checked.contains(next_node)){
                    stack.add(next_node)
                    checked.add(next_node)
                }
            }
        }
        return checked.size
    }
}